sleep/src/lib.rs

```diff
@@ -0,0 +1,224 @@
+pub struct SleepMetrics {
+    pub accelerometer: Accelerometer,
+    pub heart_rate: u8,
+    pub temperature: f64,
+    pub timestamp: std::time::Duration,
+}
+
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Accelerometer {
+    pub x: f64,
+    pub y: f64,
+    pub z: f64,
+}
+
+impl Accelerometer {
+    pub fn magnitude(&self) -> f64 {
+        (self.x.powi(2) + self.y.powi(2) + self.z.powi(2)).sqrt()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct DetectionOptions {
+    /// Allowed number of interruptions during sleeping. E.g. person rotates when sleeps and this don't mean that person woken up
+    pub allowed_magnitude_jumps: usize,
+    /// Threshold for magnitude which counts as movement
+    pub magnitude_threshold: f64,
+    /// Maximum difference for person resting heart rate and actual to consider this sleep
+    pub max_heart_rate_diff: u8,
+    /// How long there should be no movement to start tracking this as sleep
+    pub duration_for_movement: std::time::Duration,
+    /// Max delay between data points in sensors data. If delay bigger that this value sleep counting will be reset
+    pub max_delay: std::time::Duration,
+}
+
+impl DetectionOptions {
+    pub const fn new() -> Self {
+        Self {
+            allowed_magnitude_jumps: 1,
+            magnitude_threshold: 1.0,
+            max_heart_rate_diff: 5,
+            duration_for_movement: std::time::Duration::from_secs(60 * 15),
+            max_delay: std::time::Duration::from_secs(60),
+        }
+    }
+
+    pub const fn set_magnitude_threshold(mut self, magnitude_threshold: f64) -> Self {
+        self.magnitude_threshold = magnitude_threshold;
+        self
+    }
+
+    pub const fn set_duration_for_movement(
+        mut self,
+        duration_for_movement: std::time::Duration,
+    ) -> Self {
+        self.duration_for_movement = duration_for_movement;
+        self
+    }
+
+    pub const fn set_allowed_magnitude_jumps(mut self, allowed_magnitude_jumps: usize) -> Self {
+        self.allowed_magnitude_jumps = allowed_magnitude_jumps;
+        self
+    }
+
+    pub const fn set_max_heart_rate_diff(mut self, max_heart_rate_diff: u8) -> Self {
+        self.max_heart_rate_diff = max_heart_rate_diff;
+        self
+    }
+}
+
+impl Default for DetectionOptions {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug, Clone)]
+struct DataPoint {
+    pub magnitude_delta: f64,
+    pub heart_rate: (u8, u8),
+    // pub temperature: (f64, f64),
+    pub timestamp_start: std::time::Duration,
+    pub timestamp_end: std::time::Duration,
+}
+
+impl DataPoint {
+    pub fn duration(&self) -> std::time::Duration {
+        self.timestamp_end - self.timestamp_start
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum SleepPhase {
+    Deep,
+    Light,
+    REM,
+}
+
+fn sleep_detection(
```

Comment:
Some short test that will demonstrate work is good for this method. The logic behind cycles is not enough clear from first look. Consider to add it in the future, please.